### PR TITLE
CLI: Add upgrade utility with version consistency check

### PR DESCRIPTION
**Pull request by shilman**
Issue: #11382 #11383 

## What I did

```
MMBP15:cra-ts shilman$ ~/projects/baseline/storybook/lib/cli/bin/index.js upgrade --prerelease
 • Checking for latest versions of '@storybook/*' packagesinfo ,Upgrading /Users/shilman/projects/testing/cra-ts/package.json
info
info  @storybook/addon-actions  ^5.3.19  →  ^6.0.0-beta.39
info  @storybook/addon-docs     ^5.3.19  →  ^6.0.0-beta.39
info  @storybook/addon-links    ^5.3.19  →  ^6.0.0-beta.39
info  @storybook/addons         ^5.3.19  →  ^6.0.0-beta.39
info  @storybook/react          ^5.3.19  →  ^6.0.0-beta.39
info
[...]
 • Installing upgrades • Preparing to install dependencies. ✓

yarn install v1.22.4
[...]

success Saved lockfile.
✨  Done in 19.71s.
. ✓
WARN Found 1 outdated packages (relative to '@storybook/addon-actions@6.0.0-beta.39')
WARN Please make sure your packages are updated to ensure a consistent experience.
WARN - @storybook/node-logger@5.3.19
```

## How to test

```sh
cd lib/cli
yarn prepare
cd ~/my/project
~/path/to/lib/cli/bin/index.js upgrade --prerelease
```

### CODE REVIEW

File: lib/cli/src/upgrade.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    logger.warn(
+      `Found ${outdated.length} outdated packages (relative to '${formatPackage(
+        storybookPackages[0]
+      )}')`
+    );
+    logger.warn('Please make sure your packages are updated to ensure a consistent experience.');
+    warnPackages(outdated);
+  }
+
+  if (semver.gt(latestVersion, '5.4.0')) {
```
Comment (tmeasday) [inline lib/cli/src/upgrade.ts:73]: Would we have to update this magic constant every release? Should we not pull it from `package.json`?

**Review by tmeasday — CHANGES_REQUESTED** [review]
Mostly looks good, although what's up with this:

> WARN Found 1 outdated packages (relative to '@storybook/addon-actions@6.0.0-beta.39')
> WARN Please make sure your packages are updated to ensure a consistent experience.
> WARN - @storybook/node-logger@5.3.19

File: lib/cli/src/upgrade.ts (hunk 1 above)
Comment (shilman) [reply lib/cli/src/upgrade.ts]: Made this more general, but maybe not in the way you had in mind

**Comment by shilman** [comment]
> WARN Found 1 outdated packages (relative to '@storybook/addon-actions@6.0.0-beta.39')
> WARN Please make sure your packages are updated to ensure a consistent experience.
> WARN - @storybook/node-logger@5.3.19

This is a dependency of `@storybook/preset-create-react-app`. Will fix in the next version.

File: lib/cli/src/upgrade.test.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  ['│ ├─┬ @storybook/node-logger@5.3.19', { package: '@storybook/node-logger', version: '5.3.19' }],
+  [
+    'npm ERR! peer dep missing: @storybook/react@>=5.2, required by @storybook/preset-create-react-app@3.1.2',
+    null,
+  ],
+])('getStorybookVersion', (input, output) => {
+  it(input, () => {
+    expect(getStorybookVersion(input)).toEqual(output);
+  });
+});
```
Comment (ndelangen) [inline lib/cli/src/upgrade.test.ts:22]: I'm confused what this is testing

Comment (shilman) [reply lib/cli/src/upgrade.test.ts:22]: This is the output of `npm ls`

**Comment by shilman** [comment]
NOTE: I have over 100 random storybook projects on my machine (mostly user repros). I ran `sb upgrade` and `sb upgrade --prerelease` on a few of them, and found issues in each project. So I think this will generate some support issues. But more importantly, I think it will catch a lot of problems and help people clean up their Storybook installs. Really excited about it!

File: lib/cli/src/upgrade.test.ts (hunk 2 above)
Comment (ndelangen) [reply lib/cli/src/upgrade.test.ts:22]: I see.

**Review by mrmckeb — APPROVED** [review]

**Comment by ndelangen** [comment]
Perhaps some time in the future we can also automate running codemods on user's repos?

**Comment by yannbf** [comment]
Super awesome!!!!

A few ideas for the future regarding this:
1 - Mention about such command for the users [here](https://github.com/storybookjs/storybook/blob/5cec0193c6d317326f32539b918644c3256ea522/lib/core/src/server/build-dev.js#L155)
2 - Maybe add a log after the upgrades are installed with a link to migration? Could be that certain addon is updated and maybe some api changed so people would have a direction in case there's an issue after running the command
3 - Find a way to identify the need of running codemods, and if they are available then execute them
4 - Add a flag to upgrade to --next?

**Comment by shilman** [comment]
Thanks @yannbf, great suggestions. ❤️  Will address in a follow up PR for 1 -- low-hanging. 2 & 3 are both great but future work. 4 is done and it's called `--prerelease`

**Pull request merged**